cpp/tests/test_communicator.cpp

```diff
@@ -0,0 +1,136 @@
+    }
+
+    virtual MemoryType memory_type() = 0;
+
+    auto make_buffer(size_t size) {
+        if (memory_type() == MemoryType::HOST) {
+            return br->move(std::make_unique<std::vector<uint8_t>>(size));
+        } else {
+            return br->move(std::make_unique<rmm::device_buffer>(size, stream), stream);
+        }
+    }
+
+    auto copy_to_buffer(void* src, size_t size, Buffer& buf) {
+        if (memory_type() == MemoryType::HOST) {
+            std::memcpy(buf.data(), src, size);
+        } else {
+            RAPIDSMPF_CUDA_TRY(
+                cudaMemcpyAsync(buf.data(), src, size, cudaMemcpyDefault, stream)
+            );
+            RAPIDSMPF_CUDA_TRY(cudaStreamSynchronize(stream));
+        }
+    }
+
+    auto copy_from_buffer(Buffer& buf, void* dst, size_t size) {
+        if (memory_type() == MemoryType::HOST) {
+            std::memcpy(dst, buf.data(), size);
+        } else {
+            RAPIDSMPF_CUDA_TRY(
+                cudaMemcpyAsync(dst, buf.data(), size, cudaMemcpyDefault, stream)
+            );
+            RAPIDSMPF_CUDA_TRY(cudaStreamSynchronize(stream));
+        }
+    }
+
+    void log_vec(std::string const& prefix, auto const& vec) {
+        std::stringstream ss;
+        ss << prefix << " ";
+        for (auto&& v : vec) {
+            ss << static_cast<int>(v) << " ";
+        }
+        comm->logger().debug(ss.str());
+    }
+
+    Communicator* comm;
+    rmm::cuda_stream_view stream;
+    std::unique_ptr<BufferResource> br;
+};
+
+class BasicCommunicatorTest : public BaseCommunicatorTest,
+                              public testing::WithParamInterface<MemoryType> {
+  protected:
+    MemoryType memory_type() override {
+        return GetParam();
+    }
+};
+
+INSTANTIATE_TEST_CASE_P(
+    BasicCommunicatorTest,
+    BasicCommunicatorTest,
+    testing::Values(MemoryType::HOST, MemoryType::DEVICE),
+    [](const testing::TestParamInfo<MemoryType>& info) {
+        return "memory_type_" + std::to_string(static_cast<int>(info.param));
+    }
+);
+
+// Test send to self
+TEST_P(BasicCommunicatorTest, SendToSelf) {
+    if (GlobalEnvironment->type() == TestEnvironmentType::SINGLE) {
+        GTEST_SKIP() << "SINGLE communicator does not support send to self";
+    }
+
+    constexpr size_t n_elems = 10;
+    auto data = iota_vector<uint8_t>(n_elems, 0);
+    constexpr Tag tag{0, 0};
+    Rank self_rank = comm->rank();
+
+    auto send_buf = make_buffer(n_elems);
+    copy_to_buffer(data.data(), n_elems, *send_buf);
+
+    // send data to self (ignore the return future)
+    auto send_future = comm->send(std::move(send_buf), self_rank, tag);
+
+    // receive data from self
+    std::vector<std::unique_ptr<Communicator::Future>> recv_futures;
+    auto recv_buf = make_buffer(n_elems);
+    recv_futures.emplace_back(comm->recv(self_rank, tag, std::move(recv_buf)));
+
+    while (!recv_futures.empty()) {
+        auto finished = comm->test_some(recv_futures);
```

Comment:
I know when things are going down the expected path this is not necessary, I would still suggest waiting for the `send_future`, the reason this is useful is in case something goes wrong in a previous test and a receive message is waiting to be received, which is possible because we do not destroy/restart workers after each test, in that case the you may be receiving an "old" message which will be hard to debug. Additionally, would be best to check that the status of the requests are `UCS_OK`, e.g., `recv_future->getStatus() == UCS_OK` to ensure the request succeeded.

Reply:
Yes, we should check for errors, however, I suggested the "old way", instead of `getStatus()` we whould `recv_future->checkError()`, which will raise an exception if an error occurred.

Reply:
Did that in #403